src/mini_batch.jl

```diff
@@ -26,99 +33,100 @@ function kmeans!(alg::MiniBatch, X, k;
+            
             # Update per-center counts
-            N[label] += isnothing(weights) ? 1 : weights[j]  # verify (Step 11)
+            N[label] += isnothing(weights) ? 1 : weights[j]  # (Step 11)
 
             # Get per-center learning rate (Step 12)
             lr = 1 / N[label]
 
-            # Take gradient step (Step 13) # TODO: Replace with an allocation-less loop.
-            centroids[:, label] .= (1 - lr) .* centroids[:, label] .+ (lr .* batch_sample[:, j])
+            # Take gradient step (Step 13) # TODO: Replace with faster loop?
+            @views centroids[:, label] .= (1 - lr) .* centroids[:, label] .+ (lr .* X[:, j])
         end
 
-        # TODO: Reassign all labels based on new centres generated from the latest sample
-        final_labels = reassign_labels(X, metric, final_labels, centroids)
+        # Reassign all labels based on new centres generated from the latest sample
+        containers.labels .= reassign_labels(X, metric, containers.labels, centroids)
 
-        # TODO: Calculate cost on whole dataset after reassignment and check for convergence
-        J = sum_of_squares(X, final_labels, centroids)  # just a placeholder for now
+        # Calculate cost on whole dataset after reassignment and check for convergence
+        @parallelize 1 ncol sum_of_squares(containers, X, containers.labels, centroids, weights, metric)  
+        J = sum(containers.sum_of_squares)
 
         if verbose
             # Show progress and terminate if J stopped decreasing.
             println("Iteration $niters: Jclust = $J")
         end
 
-        # TODO: Check for early stopping convergence
+        # Check for early stopping convergence
         if (niters > 1) & (abs(J - J_previous) < (tol * J))
             counter += 1
 
             # Declare convergence if max_no_improvement criterion is met
             if counter >= max_no_improvement
                 converged = true
-                # TODO: Compute label assignment for the complete dataset
-                final_labels = reassign_labels(X, metric, final_labels, centroids)
+                # Compute label assignment for the complete dataset
+                containers.labels .= reassign_labels(X, metric, containers.labels, centroids)
 
-                # TODO: Compute totalcost for the complete dataset
-                J = sum_of_squares(X, final_labels, centroids)  # just a placeholder for now
+                # Compute totalcost for the complete dataset
+                @parallelize 1 ncol sum_of_squares(containers, X, containers.labels, centroids, weights, metric)
+                totalcost = sum(containers.sum_of_squares)
                 break
             end
         else
             counter = 0
+        end
 
+        # Warn users if model doesn't converge at max iterations
+        if (niters > max_iters) & (!converged)
+
+            println("Clustering model failed to converge. Labelling data with latest centroids.")
```

Comment:
No `println` in a library please :-) It may be more or less ok in other places where it is behind `verbose` flag, but here it is a warning, plain and clear and it should be showed with `@warn` macros.

Reply:
pushed to verbose in the next push